eventuals/interrupt.h

```diff
@@ -35,71 +47,97 @@ class Interrupt final {
 
     bool Install(Callback<void()>&& callback) {
       callback_ = std::move(callback);
-      return CHECK_NOTNULL(interrupt_)->Install(this);
+      return interrupt().Install(this);
     }
 
     bool Install() {
-      CHECK(callback_);
-      return CHECK_NOTNULL(interrupt_)->Install(this);
+      return interrupt().Install(this);
     }
 
     void Invoke() {
-      CHECK(callback_);
       callback_();
     }
 
     Interrupt* interrupt_ = nullptr;
     Callback<void()> callback_;
-    Handler* next_ = nullptr;
+
+    enum State : uint8_t {
+      UNINSTALLED,
+      INSTALLED,
+      UNINSTALLING,
+    };
+
+    friend std::ostream& operator<<(std::ostream& o, const State& state) {
+      if (state == UNINSTALLED) {
+        return o << "UNINSTALLED";
+      } else if (state == INSTALLED) {
+        return o << "INSTALLED";
+      } else {
+        CHECK_EQ(state, UNINSTALLING);
+        return o << "UNINSTALLING";
+      }
+    }
+
+    stout::StatefulTally<State> state_ = UNINSTALLED;
   };
 
   Interrupt()
-    : handler_(this, []() {}) {}
+    : placeholder_handler_(this, []() {}) {}
 
-  bool Install(Handler* handler) {
-    CHECK(handler->next_ == nullptr) << "handler is already installed";
+  ~Interrupt() {
+    auto* handler =
+        handler_.exchange(&placeholder_handler_, std::memory_order_relaxed);
+    if (handler != nullptr) {
+      Handler::State state = handler->state_.state();
+      handler->state_.Update(state, Handler::State::UNINSTALLED);
+    }
+  }
 
-    handler->next_ = head_.load(std::memory_order_relaxed);
+  bool Install(Handler* handler) {
+    if (handler_.load(std::memory_order_relaxed) != nullptr) {
```

Comment:
Just doing a `load()` is insufficient here because you are racing with a possible call to `Trigger()`.

You need to do a `load()`, then a `compare_exchange()`. The semantics here are that EITHER the `compare_exchange()` is successful or IT MUST be the case that `Trigger()` was called and therefore `handler_` is now set to `nullptr` so we should `CHECK()` that. That's because with our new semantics if you do a `load()` and then someone else tries to do an `Install()` then we must have a bug in our code because only one scheduler context should be installing on the same interrupt at a time.